### DISCUSSION TITLE
Transparent image rotating animation

### DISCUSSION THREAD

**Discussion by chicknsoup** [AngusJohnson/Image32#54] [General]
Hi,

How to create the rotating animation of a PNG image with TImage32? A sample demo would be great.

Thank you.

**Comment by AngusJohnson** [comment]
[Animation2.zip](https://github.com/AngusJohnson/Image32/files/15046784/Animation2.zip)

**Reply by chicknsoup** [reply]
Thank you, can we keep the background transparent?

**Reply by chicknsoup** [reply]
Another question: From your solution there should be a set of images with specific angles, and that leads to high memory usage when the angleDelta is smaller. I tried with 360 images (400x400px) and it took around 450MB of RAM. Is there any other way that we can achieve the same effect with just one or two instances of TImage32?

**Comment by AngusJohnson** [comment]
> Thank you, can we keep the background transparent?

[Animation2b.zip](https://github.com/AngusJohnson/Image32/files/15047756/Animation2b.zip)

>  I tried with 360 images (400x400px) and it took around 450MB of RAM.

Think about what you're trying to achieve! It's as if you're designing a clock with a second hand that ticks 6 times every second (ie 360 times a minute)! IOW, you almost certainly don't need so many images. 

Even if your rotation is to be really slow (which I doubt), you still wouldn't notice a rotation change of just 1 degree. And generally to avoid noticeable flicker you need about 25 frames (images) per second, though if rotation is fast you could probably get away with even fewer frames. So if, for example, you wanted one image rotation per second, then you'd only need about 25 images and a timer update every 40 msecs. The angle change between each image would be 360deg / 25 (ie 14.4 deg or 0.2513 rad). And if you're getting flicker with that, then try using 30 images/ sec instead (and redo the math with that).

**Reply by chicknsoup** [reply]
Well, the animation I'm trying to make is depending on external input ranging from -2^23 to 2^23 (which will be translated to -360 to 360 degrees). That's why the animation could be very slow.

Actually, with TImage32 from graphics32, I can easily achieve the effect with the following code. But graphics32 does not support transparency:

```
uses GR32_Transforms;

procedure Rotate(var SrcImg: TImage32; var DstImg: TImage32; Alpha: Single);
var
  W: Integer;
  H: Integer;
  T: TAffineTransformation;
begin
  W:= SrcImg.Bitmap.Width - 1;
  H:= SrcImg.Bitmap.Height - 1;
  T := TAffineTransformation.Create;
  T.SrcRect := FloatRect(0, 0, W+ 1, H+ 1);
  try
    // shift the origin
    T.Clear;

    // move the origin to a center for scaling and rotation
    T.Translate(-W* 0.5, -H* 0.5);
    T.Rotate(0, 0, Alpha);
    // move the origin back
    T.Translate(W* 0.5, H* 0.5);

    // transform the bitmap
    DstImg.BeginUpdate;
    Transform(DstImg.Bitmap, SrcImg.Bitmap, T);

    DstImg.EndUpdate;
    DstImg.Repaint;
  finally
    T.Free;
  end;
end;
```

**Reply by chicknsoup** [reply]
Btw, i'm trying to call Timage32.Rotate() multiple times within the timer (with the angle increased after each call) and my PC is running out of memory after few seconds. Is that a bug (memory leak) or it is supposed to work like that?

**Comment by AngusJohnson** [comment]
> Well, the animation I'm trying to make is depending on external input ranging from -2^23 to 2^23 (which will be translated to -360 to 360 degrees). That's why the animation could be very slow.

OK, if you really do need an image for every degree then an array of 360 rotated images is still manageable. You could store them as PNG (or JPG) resources (ie in the EXE's resource section or as a separate resource DLL). And loading these pre-rotated images from the resource section will still be _very_ much quicker than rotating them "on the fly" as you need them. And of course you'd only need one TImage32 object to do the loading and displaying.

[Animation2c.zip](https://github.com/AngusJohnson/Image32/files/15050456/Animation2c.zip)


> and my PC is running out of memory after few seconds.

You've almost certainly introduced a memory leak. (If you were just trying to do too much in the OnTimer event, then that event would sometimes be skipped.)

**Reply by chicknsoup** [reply]
Thank you so much. I have another problem with the midpoints of each rotated image being slightly different causing a bouncing side effect. Can we rotate the image around the midpoint of the original image?

**Reply by chicknsoup** [reply]
Oh, never mind. RotateRect works just great.
